1. Business Understanding
2. Data Mining
3. Data Cleaning
4. Data Exploration
5. Feature Enineering
6. Predictive Modelling
7. Data Visualisation

# Kings County House Prices
An analysis by Vivika Wilde (wilde.vivika@gmail.com).




# Objective

Determine the price drivers on house to find underpriced objects as investment oportunities.


# Set up

In [1]:
%reset -fs
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
import datetime
from scipy import stats
import collections

%matplotlib inline
#%matplotlib notebook

data = pd.read_csv(
    '/Users/vivika/nf-may-20/hh-2020-ds1-Project-EDA/King_County_House_prices_dataset.csv'
)

# Variable Names and Descriptions
from project description

* **id** - unique identified for a house
* **dateDate** - house was sold
* **pricePrice** -  is prediction target
* **bedroomsNumber** -  of Bedrooms/House
* **bathroomsNumber** -  of bathrooms/bedrooms
* **sqft_livingsquare** -  footage of the home
* **sqft_lotsquare** -  footage of the lot
* **floorsTotal** -  floors (levels) in house
* **waterfront** - House which has a view to a waterfront
* **view** - Has been viewed
* **condition** - How good the condition is ( Overall )
* **grade** - overall grade given to the housing unit, based on King County grading system
* **sqft_above** - square footage of house apart from basement
* **sqft_basement** - square footage of the basement
* **yr_built** - Built Year
* **yr_renovated** - Year when house was renovated
* **zipcode** - zip
* **lat** - Latitude coordinate
* **long** - Longitude coordinate
* **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
* **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors

# Data Types & Missings

In [2]:
kc = data.copy()
kc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
id               21597 non-null int64
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.5+ MB


In [3]:
kc.head(10)

id        date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014   221900.0         3       1.00         1180   
1  6414100192   12/9/2014   538000.0         3       2.25         2570   
2  5631500400   2/25/2015   180000.0         2       1.00          770   
3  2487200875   12/9/2014   604000.0         4       3.00         1960   
4  1954400510   2/18/2015   510000.0         3       2.00         1680   
5  7237550310   5/12/2014  1230000.0         4       4.50         5420   
6  1321400060   6/27/2014   257500.0         3       2.25         1715   
7  2008000270   1/15/2015   291850.0         3       1.50         1060   
8  2414600126   4/15/2015   229500.0         3       1.00         1780   
9  3793500160   3/12/2015   323000.0         3       2.50         1890   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0         NaN   0.0  ...      7        1180            0.0   
1      7242     2.0         0.0   0.0  ...      7        2170          400.0   
2     10000     1.0         0.0   0.0  ...      6         770            0.0   
3      5000     1.0         0.0   0.0  ...      7        1050          910.0   
4      8080     1.0         0.0   0.0  ...      8        1680            0.0   
5    101930     1.0         0.0   0.0  ...     11        3890         1530.0   
6      6819     2.0         0.0   0.0  ...      7        1715              ?   
7      9711     1.0         0.0   NaN  ...      7        1060            0.0   
8      7470     1.0         0.0   0.0  ...      7        1050          730.0   
9      6560     2.0         0.0   0.0  ...      7        1890            0.0   

  yr_built  yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0     1955           0.0    98178  47.5112 -122.257           1340        5650  
1     1951        1991.0    98125  47.7210 -122.319           1690        7639  
2     1933           NaN    98028  47.7379 -122.233           2720        8062  
3     1965           0.0    98136  47.5208 -122.393           1360        5000  
4     1987           0.0    98074  47.6168 -122.045           1800        7503  
5     2001           0.0    98053  47.6561 -122.005           4760      101930  
6     1995           0.0    98003  47.3097 -122.327           2238        6819  
7     1963           0.0    98198  47.4095 -122.315           1650        9711  
8     1960           0.0    98146  47.5123 -122.337           1780        8113  
9     2003           0.0    98038  47.3684 -122.031           2390        7570  

[10 rows x 21 columns]

In [4]:
kc.tail()

id        date     price  bedrooms  bathrooms  sqft_living  \
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
21592      1131     3.0         0.0   0.0  ...      8        1530   
21593      5813     2.0         0.0   0.0  ...      8        2310   
21594      1350     2.0         0.0   0.0  ...      7        1020   
21595      2388     2.0         NaN   0.0  ...      8        1600   
21596      1076     2.0         0.0   0.0  ...      7        1020   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
21592            0.0     2009           0.0    98103  47.6993 -122.346   
21593            0.0     2014           0.0    98146  47.5107 -122.362   
21594            0.0     2009           0.0    98144  47.5944 -122.299   
21595            0.0     2004           0.0    98027  47.5345 -122.069   
21596            0.0     2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[5 rows x 21 columns]

In [5]:
missing = pd.DataFrame(kc.isnull().sum(), columns=['Number'])
missing['Percentage'] = round(missing.Number / kc.shape[0] * 100, 1)
missing[missing.Number != 0]

Number  Percentage
waterfront      2376        11.0
view              63         0.3
yr_renovated    3842        17.8

Of the 21 variables two show incomplete data:
* waterfront is missng 11% of the values
* view is missing 0.3% of the values
* yr_renovated is missing 17.8% of the values

In [6]:
kc.waterfront.fillna(0.0, inplace=True)
kc.view.fillna(0.0, inplace=True)  #maybe mode??
kc.yr_renovated.fillna(0.0, inplace=True)

In [7]:
# years since last modernisation
kc['modernised'] = datetime.datetime.now().year - kc[[
    'yr_built', 'yr_renovated'
]].max(axis=1)
kc = kc.drop(['yr_built', 'yr_renovated'], axis=1)
kc.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      5650     1.0         0.0   0.0          3      7        1180   
1      7242     2.0         0.0   0.0          3      7        2170   
2     10000     1.0         0.0   0.0          3      6         770   
3      5000     1.0         0.0   0.0          5      7        1050   
4      8080     1.0         0.0   0.0          3      8        1680   

  sqft_basement  zipcode      lat     long  sqft_living15  sqft_lot15  \
0           0.0    98178  47.5112 -122.257           1340        5650   
1         400.0    98125  47.7210 -122.319           1690        7639   
2           0.0    98028  47.7379 -122.233           2720        8062   
3         910.0    98136  47.5208 -122.393           1360        5000   
4           0.0    98074  47.6168 -122.045           1800        7503   

   modernised  
0        65.0  
1        29.0  
2        87.0  
3        55.0  
4        33.0

In [8]:
# completing basement data
kc['sqft_basement'] = kc['sqft_living'] - kc['sqft_above']

In [9]:
kc = kc.drop(['id', 'date', 'sqft_living'], axis=1)  # removing unusable data

In [10]:
# removing outliers
outlier_indicies = list(kc[stats.zscore(kc) > 10].index)
kc.drop(outlier_indicies, inplace=True)

In [11]:
cat = kc.filter(['view', 'condition', 'grade', 'zipcode'],
                axis=1).astype("category")

cat.nunique()

view          5
condition     5
grade        11
zipcode      70
dtype: int64

In [12]:
cont = kc.drop(cat, axis=1)
cont = cont.apply(pd.to_numeric)

In [13]:
#sns.pairplot(cont)

In [14]:
cont.describe().round()

price  bedrooms  bathrooms  sqft_lot   floors  waterfront  \
count    21391.0   21391.0    21391.0   21391.0  21391.0     21391.0   
mean    530601.0       3.0        2.0   13810.0      1.0         0.0   
std     331780.0       1.0        1.0   29381.0      1.0         0.0   
min      78000.0       1.0        0.0     520.0      1.0         0.0   
25%     320000.0       3.0        2.0    5027.0      1.0         0.0   
50%     450000.0       3.0        2.0    7575.0      2.0         0.0   
75%     639000.0       4.0        2.0   10521.0      2.0         0.0   
max    4210000.0      11.0        8.0  426450.0      4.0         0.0   

       sqft_above  sqft_basement      lat     long  sqft_living15  sqft_lot15  \
count     21391.0        21391.0  21391.0  21391.0        21391.0     21391.0   
mean       1780.0          289.0     48.0   -122.0         1981.0     12031.0   
std         815.0          437.0      0.0      0.0          681.0     22360.0   
min         370.0            0.0     47.0   -123.0          399.0       651.0   
25%        1190.0            0.0     47.0   -122.0         1480.0      5100.0   
50%        1560.0            0.0     48.0   -122.0         1830.0      7600.0   
75%        2200.0          560.0     48.0   -122.0         2350.0     10004.0   
max        9410.0         4130.0     48.0   -121.0         6210.0    275299.0   

       modernised  
count     21391.0  
mean         47.0  
std          29.0  
min           5.0  
25%          21.0  
50%          43.0  
75%          66.0  
max         120.0

In [15]:
#creating dummies for categorical data
dummies = pd.DataFrame()
i_dummies = pd.DataFrame()

for i in cat:
    i_dummies = pd.get_dummies(cat[i], prefix=i, drop_first=True)
    dummies = pd.concat([dummies, i_dummies], axis=1)

kc_dum = pd.concat([cont, dummies], axis=1)
zipcodes = kc_dum.filter(regex='^zipcode', axis=1)
grades = kc_dum.filter(regex='^grade', axis=1)

In [16]:
X = list(kc_dum)
X.remove('price')
X = [x for x in X if not x.startswith('zipcode')]
X

['bedrooms',
 'bathrooms',
 'sqft_lot',
 'floors',
 'waterfront',
 'sqft_above',
 'sqft_basement',
 'lat',
 'long',
 'sqft_living15',
 'sqft_lot15',
 'modernised',
 'view_1.0',
 'view_2.0',
 'view_3.0',
 'view_4.0',
 'condition_2',
 'condition_3',
 'condition_4',
 'condition_5',
 'grade_4',
 'grade_5',
 'grade_6',
 'grade_7',
 'grade_8',
 'grade_9',
 'grade_10',
 'grade_11',
 'grade_12',
 'grade_13']

In [17]:
rs = []
adj_rs = []

for i in X:
    X1 = kc_dum[i]
    X1 = sm.add_constant(X1)
    y1 = kc_dum.price
    rs.append(sm.OLS(y1, X1).fit().rsquared.round(2))
    adj_rs.append(sm.OLS(y1, X1).fit().rsquared_adj.round(2))

X1 = zipcodes
X1 = sm.add_constant(X1)
y1 = kc_dum.price
rs.append(sm.OLS(y1, X1).fit().rsquared.round(2))
adj_rs.append(sm.OLS(y1, X1).fit().rsquared_adj.round(2))
X.append('zipcodes')

X1 = grades
X1 = sm.add_constant(X1)
y1 = kc_dum.price
rs.append(sm.OLS(y1, X1).fit().rsquared.round(2))
adj_rs.append(sm.OLS(y1, X1).fit().rsquared_adj.round(2))
X.append('grades')

rs_df = pd.DataFrame()
rs_df['explanatory_variable'] = X
rs_df['r_squared'] = rs
rs_df['adjusted_r_squared'] = adj_rs
rs_df = rs_df.sort_values('r_squared', ascending=False)
rs_df

/Users/vivika/opt/anaconda3/envs/nf/lib/python3.6/site-packages/numpy/core/fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


explanatory_variable  r_squared  adjusted_r_squared
31               grades       0.52                0.52
30             zipcodes       0.46                0.46
9         sqft_living15       0.36                0.36
5            sqft_above       0.36                0.36
1             bathrooms       0.28                0.28
27             grade_11       0.13                0.13
26             grade_10       0.13                0.13
23              grade_7       0.11                0.11
0              bedrooms       0.11                0.11
7                   lat       0.11                0.11
6         sqft_basement       0.09                0.09
25              grade_9       0.07                0.07
3                floors       0.07                0.07
28             grade_12       0.07                0.07
22              grade_6       0.05                0.05
14             view_3.0       0.04                0.04
15             view_4.0       0.04                0.04
13             view_2.0       0.03                0.03
29             grade_13       0.02                0.02
21              grade_5       0.01                0.01
2              sqft_lot       0.01                0.01
10           sqft_lot15       0.01                0.01
12             view_1.0       0.01                0.01
11           modernised       0.01                0.01
4            waterfront      -0.00               -0.00
8                  long       0.00                0.00
20              grade_4       0.00                0.00
24              grade_8       0.00                0.00
19          condition_5       0.00                0.00
18          condition_4       0.00                0.00
17          condition_3       0.00                0.00
16          condition_2       0.00                0.00

In [18]:
rs_df['explanatory_variable'].head(10)

31           grades
30         zipcodes
9     sqft_living15
5        sqft_above
1         bathrooms
27         grade_11
26         grade_10
23          grade_7
0          bedrooms
7               lat
Name: explanatory_variable, dtype: object

In [19]:
X2 = kc_dum[[
    'sqft_living15', 'sqft_above', 'bathrooms', 'bedrooms', 'sqft_basement'
]]
X2 = pd.concat([X2, zipcodes], axis=1)
X2 = sm.add_constant(X2)
y2 = kc_dum.price
sm.OLS(y2, X2).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.783
Model:                            OLS   Adj. R-squared:                  0.782
Method:                 Least Squares   F-statistic:                     1040.
Date:                Tue, 09 Jun 2020   Prob (F-statistic):               0.00
Time:                        17:02:05   Log-Likelihood:            -2.8593e+05
No. Observations:               21391   AIC:                         5.720e+05
Df Residuals:                   21316   BIC:                         5.726e+05
Df Model:                          74                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const         -1.682e+05   9654.717    -17.421      0.000   -1.87e+05   -1.49e+05
sqft_living15    66.6182      2.668     24.970      0.000      61.389      71.848
sqft_above      219.0473      2.620     83.594      0.000     213.911     224.183
bathrooms      1.042e+04   2189.273      4.759      0.000    6127.276    1.47e+04
bedrooms      -3.016e+04   1525.683    -19.767      0.000   -3.31e+04   -2.72e+04
sqft_basement   173.6966      3.160     54.975      0.000     167.504     179.890
zipcode_98002  3.334e+04   1.37e+04      2.436      0.015    6512.068    6.02e+04
zipcode_98003  6754.0841   1.23e+04      0.548      0.584   -1.74e+04    3.09e+04
zipcode_98004  7.921e+05   1.21e+04     65.693      0.000    7.68e+05    8.16e+05
zipcode_98005  3.389e+05   1.45e+04     23.314      0.000     3.1e+05    3.67e+05
zipcode_98006  3.203e+05   1.09e+04     29.410      0.000    2.99e+05    3.42e+05
zipcode_98007  2.718e+05   1.54e+04     17.661      0.000    2.42e+05    3.02e+05
zipcode_98008  2.757e+05   1.24e+04     22.192      0.000    2.51e+05       3e+05
zipcode_98010  7.376e+04   1.76e+04      4.197      0.000    3.93e+04    1.08e+05
zipcode_98011  1.121e+05   1.38e+04      8.134      0.000    8.51e+04    1.39e+05
zipcode_98014  1.022e+05   1.65e+04      6.176      0.000    6.97e+04    1.35e+05
zipcode_98019  5.956e+04    1.4e+04      4.258      0.000    3.21e+04     8.7e+04
zipcode_98022  3.761e+04   1.32e+04      2.839      0.005    1.16e+04    6.36e+04
zipcode_98023 -1.896e+04   1.07e+04     -1.769      0.077      -4e+04    2044.754
zipcode_98024  1.717e+05   1.93e+04      8.873      0.000    1.34e+05     2.1e+05
zipcode_98027  1.707e+05   1.12e+04     15.186      0.000    1.49e+05    1.93e+05
zipcode_98028  1.216e+05   1.23e+04      9.876      0.000    9.75e+04    1.46e+05
zipcode_98029  2.145e+05   1.19e+04     17.978      0.000    1.91e+05    2.38e+05
zipcode_98030 -3148.0129   1.27e+04     -0.249      0.804   -2.79e+04    2.17e+04
zipcode_98031  1.193e+04   1.24e+04      0.961      0.337   -1.24e+04    3.63e+04
zipcode_98032   2.59e+04   1.61e+04      1.610      0.107   -5621.877    5.74e+04
zipcode_98033  3.854e+05   1.11e+04     34.777      0.000    3.64e+05    4.07e+05
zipcode_98034  2.112e+05   1.05e+04     20.053      0.000    1.91e+05    2.32e+05
zipcode_98038  5967.9955   1.04e+04      0.575      0.566   -1.44e+04    2.63e+04
zipcode_98039  1.255e+06   2.41e+04     52.005      0.000    1.21e+06     1.3e+06
zipcode_98040  5.524e+05   1.26e+04     43.804      0.000    5.28e+05    5.77e+05
zipcode_98042  1597.9046   1.05e+04      0.152      0.879    -1.9e+04    2.22e+04
zipcode_98045  1.086e+05   1.33e+04      8.143      0.000    8.25e+04    1.35e+05
zipcode_98052  2.322e+05   1.05e+04     22.210      0.000    2.12e+05    2.53e+05
zipcode_98053  1.797e+05   1.13e+04     15.845   

In [20]:
X3 = kc_dum.copy()
X3.drop('price', axis=1, inplace=True)
X3 = sm.add_constant(X3)
y3 = kc_dum.price
model = sm.OLS(y3, X3).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.837
Model:                            OLS   Adj. R-squared:                  0.836
Method:                 Least Squares   F-statistic:                     1112.
Date:                Tue, 09 Jun 2020   Prob (F-statistic):               0.00
Time:                        17:02:06   Log-Likelihood:            -2.8290e+05
No. Observations:               21391   AIC:                         5.660e+05
Df Residuals:                   21292   BIC:                         5.668e+05
Df Model:                          98                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const         -3.196e+07   5.19e+06     -6.152      0.000   -4.21e+07   -2.18e+07
bedrooms      -8153.7759   1386.917     -5.879      0.000   -1.09e+04   -5435.315
bathrooms      2.303e+04   2239.644     10.281      0.000    1.86e+04    2.74e+04
sqft_lot          0.3648      0.053      6.940      0.000       0.262       0.468
floors        -3.022e+04   2682.015    -11.267      0.000   -3.55e+04    -2.5e+04
waterfront        0.0309      0.005      6.151      0.000       0.021       0.041
sqft_above      148.7172      2.692     55.246      0.000     143.441     153.994
sqft_basement    96.7647      3.053     31.693      0.000      90.780     102.749
lat            1.587e+05   5.37e+04      2.958      0.003    5.35e+04    2.64e+05
long              -2e+05   3.86e+04     -5.187      0.000   -2.76e+05   -1.24e+05
sqft_living15    23.2125      2.478      9.366      0.000      18.355      28.070
sqft_lot15       -0.1827      0.071     -2.591      0.010      -0.321      -0.045
modernised      108.7077     53.414      2.035      0.042       4.013     213.403
view_1.0       9.332e+04   7672.470     12.163      0.000    7.83e+04    1.08e+05
view_2.0        7.41e+04   4708.615     15.738      0.000    6.49e+04    8.33e+04
view_3.0       1.644e+05   6514.370     25.239      0.000    1.52e+05    1.77e+05
view_4.0        3.08e+05   1.01e+04     30.385      0.000    2.88e+05    3.28e+05
condition_2     6.95e+04   2.76e+04      2.516      0.012    1.54e+04    1.24e+05
condition_3    8.756e+04   2.57e+04      3.405      0.001    3.72e+04    1.38e+05
condition_4    1.124e+05   2.57e+04      4.367      0.000    6.19e+04    1.63e+05
condition_5    1.595e+05   2.59e+04      6.164      0.000    1.09e+05     2.1e+05
grade_4       -1.178e+05   1.37e+05     -0.858      0.391   -3.87e+05    1.51e+05
grade_5       -1.423e+05   1.35e+05     -1.053      0.292   -4.07e+05    1.22e+05
grade_6       -1.544e+05   1.35e+05     -1.145      0.252   -4.19e+05     1.1e+05
grade_7       -1.493e+05   1.35e+05     -1.107      0.268   -4.14e+05    1.15e+05
grade_8       -1.234e+05   1.35e+05     -0.915      0.360   -3.88e+05    1.41e+05
grade_9       -4.373e+04   1.35e+05     -0.324      0.746   -3.08e+05    2.21e+05
grade_10       7.458e+04   1.35e+05      0.552      0.581    -1.9e+05    3.39e+05
grade_11       2.614e+05   1.35e+05      1.933      0.053   -3617.387    5.26e+05
grade_12       5.805e+05   1.36e+05      4.267      0.000    3.14e+05    8.47e+05
grade_13        1.24e+06   1.42e+05      8.738      0.000    9.62e+05    1.52e+06
zipcode_98002  2.054e+04   1.21e+04      1.695      0.090   -3216.968    4.43e+04
zipcode_98003 -1.295e+04   1.08e+04     -1.196      0.232   -3.42e+04    8278.818
zipcode_98004  7.205e+05   1.98e+04     36.297      0.000    6.82e+05    7.59e+05
zipcode_98005  2.868e+05   2.12e+04     13.536   

In [21]:
X3['valued_at'] = model.predict(X3).round(2)
X3.info()

#result = pd.merge(X3, kc[['id', 'date', 'yr_built','yr_renovated', 'sqft_living']],
#                 on='use_id',
#                 how='left')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21391 entries, 0 to 21596
Columns: 101 entries, const to valued_at
dtypes: float64(8), int64(6), uint8(87)
memory usage: 4.2 MB


In [23]:
result = data.join(X3['valued_at'])
result['valuation_deviation_abs'] = (result['valued_at'] - result['price']).round(2)
result['valuation_deviation_pct'] = (result['valuation_deviation_abs']/result['price']*100).round(2)
result.sort_values(by = 'valuation_deviation_pct').head(10)

id        date     price  bedrooms  bathrooms  sqft_living  \
15698  3089000005   7/24/2014  150000.0         2       1.00          850   
13744  1788900230    4/3/2015  199950.0         3       1.00          840   
5719   1788800630   2/25/2015  185000.0         3       1.00          840   
4199   1437500015    7/9/2014  150000.0         3       0.75          490   
19581  1788800770   7/28/2014  187500.0         3       1.00          840   
13616  2023049218   3/16/2015  445000.0         2       1.00          930   
3329   3410600015    4/6/2015  250000.0         2       0.75          700   
5254   8651431100   1/16/2015  199990.0         3       1.00          840   
11677  3121059036  10/29/2014  400000.0         2       1.00         1140   
5416   1865810290   3/18/2015  232000.0         3       1.00          840   

       sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  \
15698     54000     1.5         0.0   0.0  ...      1950           0.0   
13744      9480     1.0         0.0   0.0  ...      1960           0.0   
5719      12091     1.0         0.0   0.0  ...      1959           0.0   
4199      38500     1.5         0.0   0.0  ...      1959           0.0   
19581      8400     1.0         0.0   0.0  ...      1959           0.0   
13616      7740     1.0         0.0   0.0  ...      1932           NaN   
3329      16828     1.0         0.0   0.0  ...      1958           0.0   
5254       5200     1.0         0.0   0.0  ...      1969        2014.0   
11677    101529     1.5         0.0   0.0  ...      1932           0.0   
5416       6540     1.0         0.0   0.0  ...      1970        2014.0   

       zipcode      lat     long  sqft_living15  sqft_lot15  valued_at  \
15698    98023  47.2959 -122.377           1550       14440   19054.18   
13744    98023  47.3277 -122.341            840        9420   40434.05   
5719     98023  47.3281 -122.343            840        9324   41976.33   
4199     98014  47.7112 -121.315            800       18297   36041.02   
19581    98023  47.3281 -122.344           1030        8640   45365.33   
13616    98148  47.4611 -122.324           1620        8584  109019.21   
3329     98092  47.3009 -122.125           2010       29316   63944.02   
5254     98042  47.3685 -122.077            870        5200   58043.61   
11677    98092  47.2592 -122.228           1580      101529  119765.60   
5416     98042  47.3743 -122.116           1010        6600   70248.76   

       valuation_deviation_abs  valuation_deviation_pct  
15698               -130945.82                   -87.30  
13744               -159515.95                   -79.78  
5719                -143023.67                   -77.31  
4199                -113958.98                   -75.97  
19581               -142134.67                   -75.81  
13616               -335980.79                   -75.50  
3329                -186055.98                   -74.42  
5254                -141946.39                   -70.98  
11677               -280234.40                   -70.06  
5416                -161751.24                   -69.72  

[10 rows x 24 columns]